### PR TITLE
HDDS-5883 Change S3G client to set S3 Auth per req

### PR DESCRIPTION
**Pull request by kerneltime**
## What changes were proposed in this pull request?

S3G client will now set the S3 Auth information for
each OM Request. S3G itself will authenticate as S3G
identity registered with Kerberos but Request
Processing should be done via the S3 Auth set in the
protobuf request.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-5883

## How was this patch tested?

This was tested with code for 
https://issues.apache.org/jira/browse/HDDS-5884 and
https://issues.apache.org/jira/browse/HDDS-5885
Objects created were inspected in the OM DB. Objects were created using OM's key tab 
``` "fileName": "key2",
  "acls": [
    {
      "type": "USER",
      "name": "om/om@EXAMPLE.COM",
      "aclBitSet": [
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        1
      ],
      "aclScope": "ACCESS"
    },
    {
      "type": "GROUP",
      "name": "root",
      "aclBitSet": [
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        1
      ],
      "aclScope": "ACCESS"
    }
  ],
  "parentObjectID": 0,
  "objectID": -9223372036854773504,
  "updateID": 11,
  "metadata": {}
}
```

### CODE REVIEW

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocolPB/OzoneManagerProtocolClientSideTranslatorPB.java [hunk 1]
```diff
@@ -223,11 +224,24 @@ public void close() throws IOException {
...
-    OMRequest payload = OMRequest.newBuilder(omRequest)
-        .setTraceID(TracingUtil.exportCurrentSpan())
-        .build();
-
-    return transport.submitRequest(payload);
+    OMRequest.Builder  b = OMRequest.newBuilder(omRequest);
+    // Insert S3 Authentication information for each request.
+    if (getThreadLocalS3Authentication() != null) {
+      b.setS3Authentication(
+          org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.
```
[the remark this answers is not in the document]
Comment (kerneltime) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocolPB/OzoneManagerProtocolClientSideTranslatorPB.java]: Renamed it.

**Review by avijayanhwx — COMMENTED** [review]
Thanks for working on this @kerneltime. I am +1 on this change.

Just something to keep in mind, we need to make sure the error codes for s3 request failures (wrong creds, OM down etc) are semantically correct and consistent in both the RPC and GRPC implementations. For example, we now (incorrectly) return a 500 error for wrong credentials.

**Comment by kerneltime** [comment]
> Thanks for working on this @kerneltime. I am +1 on this change.
> 
> Just something to keep in mind, we need to make sure the error codes for s3 request failures (wrong creds, OM down etc) are semantically correct and consistent in both the RPC and GRPC implementations. For example, we now (incorrectly) return a 500 error for wrong credentials.

Thank you for bringing it up, have filed https://issues.apache.org/jira/browse/HDDS-5918 to track it

**Comment by bharatviswa504** [comment]
Thank You @avijayanhwx for the review and @kerneltime for the contribution

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/OzoneClientProducer.java [hunk 2]
```diff
@@ -77,23 +71,22 @@
   private ContainerRequestContext context;
 
   @Produces
-  public OzoneClient createClient() throws WebApplicationException,
+  public synchronized OzoneClient createClient() throws WebApplicationException,
```
Comment (bharatviswa504) [inline hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/OzoneClientProducer.java:74]: We donot need to throw WebApplicationException. Discussed offline which will be done in the next PR.

**Pull request merged**